### DISCUSSION TITLE
How to add new physical properties to a fluid

### DISCUSSION THREAD

**Discussion by shin-chan-cpu** [idaholab/moose#27151] [Q&A Modules: Navier-Stokes]
Hello, I am a newbie who just started using the ns module. I want to simulate the flow of a fluid heated by a pipe in moose. I want my density, dynamic viscosity, specific heat capacity and thermal conductivity to change with temperature. How can I achieve such a simulation in moose?

**Comment by GiudGiud** [comment]
Hello

There are very many options for fluid properties:
I think you are looking for this one
https://mooseframework.inl.gov/source/fluidproperties/TemperaturePressureFunctionFluidProperties.html

Guillaume

**Reply by shin-chan-cpu** [reply]
Thank you very much for your suggestion, but after I wrote the temperature in the fluid to the input card like yours, he told me that the variable type could not be converted, and the error was displayed
`'No postprocessor, scalar variable, or function with the name 'T_fluid' found. Unable to convert 'T_fluid' to type double`
Is this because I set the T_fluid this variable type is INSFVEnergyVariable?

**Reply by GiudGiud** [reply]
The example is not great. Write down T_fluid in the expression as x, and pressure as y.

**Reply by shin-chan-cpu** [reply]
It says like this in my input card, but he still has problems.
```
  [rho]
    type = ParsedFunction
    symbol_names = 'T p'
    symbol_values = 'T_fluid pressure'
    expression = '1.5e3 + 0.13 * T - 1.5e-4 * p'
  []
```
Should I use some auxkernel to convert it?

**Reply by GiudGiud** [reply]
No. Use x and y like I mentioned

you do not need to use symbol names and vars if you use x and y

self reminder for me to fix the example to be clearer

**Reply by GiudGiud** [reply]
```
  [rho]
    type = ParsedFunction
    expression = '1.5e3 + 0.13 * x - 1.5e-4 * y'
  []
```

**Reply by shin-chan-cpu** [reply]
But does this vary with temperature?

**Reply by GiudGiud** [reply]
Yes it does. X gets substituted with temperature

**Reply by shin-chan-cpu** [reply]
Thank you very much. I'll give it a try.

**Reply by shin-chan-cpu** [reply]
After I wrote it out according to the input card, I reported the following error:
```
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Attempted to get a functor with the name 'k' but multiple functors match. Make sure that you do not have functor material properties, functions, postprocessors or variables with the same names
```
But my input card only shows k in these places.
```
[Functions]
  [k]
   type = ParsedFunction
    expression = '0.08 '
  []
  [rho]
    type = ParsedFunction
    expression = '1.5e3 + 0.13 * x'
  []
  [mu]
    type = ParsedFunction
    expression = '1e-3 + 2e-6 * x'
  []
[]

[FluidProperties]
  [fp]
    type = TemperaturePressureFunctionFluidProperties
    cv = ${cv}
    k = k
    rho = rho
    mu = mu
  []
[]
[Materials]
  [to_vars]
    type = FluidPropertiesMaterialPT
    fp = fp
    outputs = 'all'
    output_properties = 'density k cp cv viscosity e h'
    pressure = pressure
    temperature = T_fluid

    compute_entropy = false
    compute_sound_speed = false
  []
[]
```
